### ISSUE TITLE
Flatpak package

### ISSUE THREAD

**Issue by AdrianKoshka** [debauchee/barrier#30]
I was wondering if there'd be any interest in packaging barrier in flatpak format. Distros like [Fedora Atomic Workstation](https://www.projectatomic.io/blog/2018/02/fedora-atomic-workstation/) encourage the use of flatpaks for GUI applications over traditionally packaged applications. I help maintain the Thunderbird flatpak so I'd be willing to help package barrier. Getting it up on flathub would also help.

**Comment by walker0643** [comment]
hi @AdrianKoshka - of course we would have no objection to that. if you're offering to maintain the package then just let us know what you need from us to get started. cheers!

**Comment by AdrianKoshka** [comment]
A list of dependencies that barrier needs to compile against would be a good start, and then how to build in on linux. I see there's an issue filed already for build instructions for linux (#28).

**Comment by AdrianKoshka** [comment]
Seems avahi libraries aren't part of the KDE/FreeDesktop runtime, what version of avahi is barrier 2.0.0 built against?

**Comment by AdrianKoshka** [comment]
If you want to keep up with the repo: https://github.com/AdrianKoshka/flathub/tree/com.gituhb.debauchee.barrier

**Comment by AdrianKoshka** [comment]
> Seems avahi libraries aren't part of the KDE/FreeDesktop runtime, what version of avahi is barrier 2.0.0 built against?

In what capacity does barrier make use of avahi? I ask because I realize if I have to compile avahi daemon and all, it would just be more efficient to (in my opinion):

- Ask flatpak to add Avahi protal
- Only build dev libraries so barrier can at least build
- Go from there?

**Comment by walker0643** [comment]
avahi is used to do auto configuration (auto-connect) between server and client on the LAN. I've been linking it against 0.7 on linux platforms.

as a quick rundown, barrier also needs the following to build: CMake, Python, GNU Make, GCC v4.9+, X11 headers and libs, Qt 5 (core, ui, widgets, network, qmake), libcurl, and openssl headers and libraries.

there is a clean_build.sh script in the base directory to automate the process a little. anything you need to tweak in the environment should go into a new script beside it called build_env.sh (make it +x). git won't track it nor overwrite it when you pull (it's a site config).

**Comment by AdrianKoshka** [comment]
Thanks for the info.

**Comment by AdrianKoshka** [comment]
Do you know anything about building avahi?

**Comment by walker0643** [comment]
I've not built it yet; so far the target platforms have all had usable packages. Are you running into trouble?

**Comment by AdrianKoshka** [comment]
Yeah...the build options are just poorly documented or just not documented at all. I've disabled quite a bit but can't get everything I need disabled (like the daemon component).

**Comment by AdrianKoshka** [comment]
Thanks to @lathiat, I was able to get avahi to at least compile. Though I still seem to be getting this error:

```
-- Looking for include file dns_sd.h
-- Looking for include file dns_sd.h - not found
CMake Error at CMakeLists.txt:191 (message):
  Missing header: dns_sd.h


-- Configuring incomplete, errors occurred!
See also "/run/build/barrier/CMakeFiles/CMakeOutput.log".
See also "/run/build/barrier/CMakeFiles/CMakeError.log".
Error: module barrier: Child process exited with code 1
make: *** [Makefile:8: build] Error 1
```

**Comment by AdrianKoshka** [comment]
So If I remember correctly the default config file is `~/.barrier.config`? right? I was wondering if instead you could use `XDG_CONFIG_HOME`. (so something like `~/.config/barrier/<files-within>`)

**Comment by AdrianKoshka** [comment]
Also use `XDG_DATA_HOME` to store the SSL certs?

**Comment by AdrianKoshka** [comment]
http://docs.flatpak.org/en/latest/conventions.html#xdg-base-directories

**Comment by AdrianKoshka** [comment]
Also, apparently the `.desktop` file's executable path is hardocded. An excerpt of a conversation I had with a flatpak dev on what's wrong with the `.desktop` file you supply with barrier:

```
TingPing hardcoding /usr/bin is wrong any time you have a different prefix
TingPing either A) Use just the binary name, no path. B) At configure time insert the correct prefix
```

**Comment by AdrianKoshka** [comment]
Also, there seems to be something wrong the `include` check you have CMake do. As of now, I have to manually tell it where to look for the `dns_sd.h` header for avahi in the flatpak manifest. An excerpt from the manifest:

```json
"build-options": {
    "cflags": "-O2 -g -I/app/include/avahi-compat-libdns_sd",
    "cxxflags": "-O2 -g -I/app/include/avahi-compat-libdns_sd",
    "env": {
      "V": "1"
    }
  },
```

**Comment by walker0643** [comment]
setting BONJOUR_SDK_HOME in the environment should take care of the non-standard location of the avahi package. eg. if the header is /opt/bonjour/include/dns_sd.h and the lib is /opt/bonjour/lib/dnssd.lib then add "export BONJOUR_SDK_HOME=/opt/bonjour" to your build_env.sh file in the root of the source.

**Comment by walker0643** [comment]
as far as the config file locations, can you create a new issue for that so it can be addressed separately?

**Comment by AdrianKoshka** [comment]
Can do.

**Comment by AdrianKoshka** [comment]
Issue #31 created.

**Comment by AdrianKoshka** [comment]
> then add "export BONJOUR_SDK_HOME=/opt/bonjour" to your build_env.sh file in the root of the source.

The build scripts are actually bipassed entirely when I build with flatpak, I think flatpak-build just knows magically how to run Cmake and friends.

**Comment by AdrianKoshka** [comment]
```json
"build-options": {
        "env": {
          "BARRIER_REVISION": "2.0.0",
	      "BONJOUR_SDK_HOME": "/app/include"
        }
      },
```

**Comment by AdrianKoshka** [comment]
Setting it via:

```json
  "build-options": {
    "cflags": "-O2 -g",
    "cxxflags": "-O2 -g",
    "env": {
      "V": "1",
      "BONJOUR_SDK_HOME": "/app/include"
```
doesn't work either...

**Comment by AdrianKoshka** [comment]
The contents of `app/include` btw:

```
total 0
drwxrwxr-x. 2 alc alc  55 Apr 26 18:49 avahi-client
drwxrwxr-x. 2 alc alc 260 Apr 26 18:49 avahi-common
drwxrwxr-x. 2 alc alc  22 Apr 26 18:49 avahi-compat-libdns_sd
drwxrwxr-x. 2 alc alc  78 Apr 26 18:49 avahi-core
```

**Comment by AdrianKoshka** [comment]
This appears to be a CMake issue?

because:

1. The env variable works just fine in a shell module
2. I was told so by somebody helping me in `#flatpak`

**Comment by walker0643** [comment]
oops that was my bad... BONJOUR_SDK_HOME only works on windows. sorry :stuck_out_tongue:

yes, if cmake isn't finding your bonjour libs then just add your include and link options as you were doing. it's certainly cleaner to find out why cmake isn't finding bonjour and fix that instead, but for now just getting through the build might be enough.

btw, might it have something to do with the fact that bonjour is under /app ?? i'm not familiar with flatpak but /app seems like a very non-standard location. is everything used by the build process under /app or just bonjour?

**Comment by AdrianKoshka** [comment]
Everything, I can get you a link of flatpak file-structure if you want. While I would rather not do that include thing, but if it makes it work for now and will actually get solved later, I'll do it. I haven't even gotten around to see if barrier launches/work yet, but will probably later today.

here's that [link](http://docs.flatpak.org/en/latest/conventions.html#filesystem-layout).

**Comment by AdrianKoshka** [comment]
I actually probably won't test that it works until #31 is solved, I don't feel like opening the sandbox.

**Comment by AdrianKoshka** [comment]
I lied, and I can confirm it at least launches. Oh, question, does Barrier need DRI access? I ask because I see this:

```
alc@am1m-s2h ~/git/github/flathub/flathub (git)-[com.gituhb.debauchee.barrier] % flatpak run com.github.debauchee.barrier
libGL error: MESA-LOADER: failed to retrieve device information
libGL error: unable to load driver: amdgpu_dri.so
libGL error: driver pointer missing
libGL error: failed to load driver: amdgpu
libGL error: failed to open drm device: No such file or directory
libGL error: failed to load driver: radeonsi
```

**Comment by AdrianKoshka** [comment]
![picture of barrier running via flatpak run](https://imgur.com/4yno90B.png)

**Comment by AdrianKoshka** [comment]
Did a single test, and the keyboard/mouse sharing worked, as well as copy/paste:

- Host: Fedora 27, flatpak'd `barrier` (running without SSL) 2.0.0
- Client: Debian 9 stretch, `~/.local/bin/barrierc --no-daemon <server-ip>` 2.0.0

**Comment by AdrianKoshka** [comment]
[Here's a rough-draft of the appstream metadta](https://github.com/AdrianKoshka/flathub/blob/com.github.debauchee.barrier/com.github.debauchee.barrier.appdata.xml)

**Comment by AdrianKoshka** [comment]
I built barrier as a 32-bit flatpak, and it runs!

![32-bit barrier flatpak running](https://imgur.com/c0NWQVu.png)

**Comment by AdrianKoshka** [comment]
Here's the 32-bit flatpak'd barrier running on 32-bit `flatpak` on my netbook:

![32-bit flatpak netbook](https://imgur.com/wWajhsM.png)

**Comment by AdrianKoshka** [comment]
Apologies for the radio silence, I was on a vacation for the past two weeks, and went to VCF-E (Vintage Computer Festival East) this past weekend.

**Comment by AdrianKoshka** [comment]
Thanks for b43581c2f54ebb38eefe635352a8ae4ac980407e, I was able to [remove a section from my manifest](https://github.com/AdrianKoshka/flathub/commit/e74fd81b9cc27a31d3e067077e6766e8253316a7).

**Comment by johnny-mac** [comment]
So this is in active development on flatpak? that's awesome! But your link up there didn't work, page not found.. You gonna be adding it to flathub any time soon? 
I run Void Linux (GLIBC & MUSL) but can't get Barrier to build, would really love to be able to use a flatpak app for it.

**Comment by AdrianKoshka** [comment]
I'm going to PR to flathub when I feel it's ready. I already have one WIP PR to flathub (for virt-viewer/remote-viewer) so I don't have to WIP PR's.

**Comment by AdrianKoshka** [comment]
So far, the easiest way to build the flatpak is:

```shell
$ flatpak --user install flathub org.flatpak.Builder
$ git clone git@github.com:AdrianKoshka/flathub.git -b com.github.debauchee.barrier
$ cd flathub
$ make FLATPAK_BUILDER="flatpak run org.flatpak.Builder
```
Then to install locally:

```shell
$ flatpak --user remote-add lbr /full/path/to/repo/
$ flatpak --user install lbr com.github.debauchee.barrier
```

**Comment by AdrianKoshka** [comment]
Why was this closed? Inactivity, I've been waiting on a PR you pulled into a different branch to be pulled into master.

**Comment by AdrianKoshka** [comment]
Now that the `pkg-config` fixed was merged, I can say that I no longer have to use the hack to point the compiler to the avahi libraries.

**Comment by AdrianKoshka** [comment]
This means I can finally start moving forward again, probably start working on the appstream data.

**Comment by AdrianKoshka** [comment]
How do you properly pass the version of barrier to the buildsystem? I have:

```json
{
            "name" : "barrier",
            "buildsystem" : "cmake-ninja",
            "build-options" : {
                "env" : {
                    "BARRIER_REVISION" : "2.1.2"
                }
            },
            "sources" : [
                {
                    "type" : "git",
                    "url" : "git://github.com/debauchee/barrier.git",
                    "tag" : "v2.1.2",
                    "commit" : "cc69299ea39bd39caad54ae349564e22593eb4df"
                }
```

Though when I run `--version`, I get:

```
alc@am1m-s2h ~ % flatpak run --command=barrierc com.github.debauchee.barrier --version
barrierc 2.1.0-snapshot
Protocol version 1.6
Copyright (C) 2018 Debauchee Open Source Group
Copyright (C) 2012-2016 Symless Ltd.
Copyright (C) 2008-2014 Nick Bolton
Copyright (C) 2002-2014 Chris Schoeneman
```

**Comment by walker0643** [comment]
The top of cmake/Version.cmake has the constants that cmake uses to fill in the placeholder variables like @BARRIER_VERSION@ in the .in files. Is that what you were asking??

**Comment by ugtar** [comment]
is avahi required? i personally don't have the avahi daemon running because i don't need to auto-discover anything.  With avahi not running, barrier can't auto-configure, but it works fine if I manually enter the barrier server IP.  Seems like avahi could be an optional dependency that just disables the auto-discovery option at build time.

**Comment by AdrianKoshka** [comment]
I'm not going to take avahi out just because one user doesn't use it. Also I should've closed this issue a while ago.

**Comment by AdrianKoshka** [comment]
@ugtar If your intention is to make avahi an optional build component (which it [seems to be, partially](https://github.com/debauchee/barrier/blob/master/CMakeLists.txt#L199-L201)), then open another issue instead of piggy-backing off of this one.
